Docs: Add Parameters API reference

kylegach opened the pull request:
Contributes to (but does not close): https://github.com/storybookjs/storybook/issues/17243

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

<!-- Briefly describe what your PR does -->

- Add a Parameters API reference
- There are very few parameters for Storybook itself, so most of the work is adding API sections to the essential addons docs

See also: https://github.com/storybookjs/addon-kit/pull/61

<img width="691" alt="" src="https://github.com/storybookjs/storybook/assets/486540/688556b6-b11f-40b3-8ec8-b8fc51232435">

## Checklist for Contributors

### Testing

#### Manual testing

1. Follow the steps in the [contributing instructions](https://storybook.js.org/docs/react/contribute/new-snippets#preview-your-work) for this branch, `api-ref-parameters`

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [x] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

kylegach committed:
Add Parameters API reference

- Add API references for all essential addons
- Consolidate related snippets

shilman commented on docs/api/parameters.md:
````diff
@@ -0,0 +1,238 @@
+}
+```
+
+#### `options.storySort`
````
Note: This is only valid in project annotations. It probably should not be a parameter, but we haven't changed it yet

shilman commented on docs/api/parameters.md:
```diff
@@ -0,0 +1,238 @@
+
+<div class="aside">
+
+ℹ️ Parameters are **merged**, so individual keys are always overwritten, never dropped.
```
Probably worth clarifying the merge semantics in detail here. Objects are deep-merged, but arrays are replaced.

So merging 

```
{
  a: 'a1',
  nested: { b: 'b1' },
  array: [1, 2]
}
```

With 

```
{
  a: 'a2',
  nested: { c: 'c2' },
  array: [3, 4]
}
```

would yield

```
{
  a: 'a2',
  nested: { b: 'b1', c: 'c2' },
  array: [3, 4]
}
```

shilman reviewed (approved):
Great work @kylegach !! A few minor points

kylegach committed:
Address feedback

- Clarify that `options` can only be applied at project level
- Make parameters merge semantics more clear
